server/routerlicious/packages/routerlicious-base/src/riddler/tenantManager.ts

```diff
@@ -76,6 +79,10 @@ export class TenantManager {
             // if the tenant doesn't have key2, it will be empty string
             // we should fail token generated with empty string as key
             if (!tenantKeys.key2) {
+                Lumberjack.info(`Invalid key. Deleting from cache.`, lumberProperties);
+                this.cache.delete(tenantId).catch((error) => {
```

Comment:
My understanding is we should only delete the key from cache on refreshTenantKey, or on some TTL. Just because a users JWT is expired or invalid, it does not mean that the key in our cache is invalid.

In other words, the validity of a user-provided token has no relation to the validity of the key that we are verifying the token against. We should remove the cache deletes in this method.

Reply:
I am wondering if there could be cases like user initiated token revocation or other scenarios where the value in cache might end up being stale. Having a fallback to db in these cases sounds like a useful hedge against these edge cases at the cost of additional db calls.

Reply:
Removed cache deletes from this method

Reply:
@znewton Considering what @pradeepvairamani said, would we then want to still delete from the cache when the token is invalid? The next time the getTenantKeys is called we would read from the database first then add the value to the cache. It would be an additional call to the database, but could help in cases like those listed above

Reply:
If the concern is that `this.getTenantKeys` is returning stale information, then I think the problem needs to be addressed there. Deleting the cached key on a 403 won't help the user recover, because the 403 is not retriable, and deleting the cached key on a 401 isn't helpful because that's just saying the key is valid but the token's expiration date is past.

My understanding of token revocation is for specific JWTs, _not keys_, which is a totally different piece of work. Are you talking about `refreshTenantKey`? In that case, this PR already includes cache invalidation on key refresh. The only problem I can see there would be if the session lands on a different cluster than where the refreshTenantKey landed, in which case a user with the **old** key would get a 200, and a user with the **new** key would get a 403, both of which are incorrect, but neither of which are solved by clearing cache on a 403/401.

I think that's a totally valid concern, but needs a different solution than clearing cache every time we block a user with an invalid token.

Reply:
The scenario that needs to be addressed is if we have 2 clusters, ClusterA and ClusterB, User, and KeyA. Scenario: UserA is working on a document in ClusterA using a token signed with KeyA, then C1 initiates a tokenRefresh that lands on ClusterB and switches the valid tenant key to KeyB. 

I think the more complete solution that is needed would be clear cache on a 403 (when using a cached key) then retry token validation before throwing a 403 to the user.

401s shouldn't clear cache, IMO

Reply:
So then the code would look something like

```ts

    /**
     * Validates a tenant's API token
     */
    public async validateToken(tenantId: string, token: string, includeDisabledTenant = false): Promise<void> {
        const lumberProperties = { [BaseTelemetryProperties.tenantId]: tenantId };
        // Alter return type to include `fromCache`, or make some other change that allows us to know that the retried verification is with a key from DB
        const { keys, fromCache } = await this.getTenantKeys(tenantId, includeDisabledTenant);

        return jwt.verify(token, tenantKeys.key1, (error1) => {
            if (!error1) {
                return;
            }
            // if the tenant doesn't have key2, it will be empty string
            // we should fail token generated with empty string as key
            if (!tenantKeys.key2) {
                Lumberjack.error(`Error with tenant key 1 token verification.`, lumberProperties, error1);
                const error = error1 instanceof jwt.TokenExpiredError
                    ? new NetworkError(401, "Token expired validated with key1.")
                    : new NetworkError(403, "Invalid token validated with key1.");
                // IMPORTANT to know if the key used for verification was cached, otherwise this will be an infinite loop when the key is actually invalid
                if (error.code === 403 && fromCache) {
                   await this.cache.delete(this.getCacheKey(tenantId));
                   return this.validateToken(tenantId, token, includeDisabledTenant);
                } else {
                   throw error;
                }
            }
            jwt.verify(token, tenantKeys.key2, (error2) => {
                if (!error2) {
                    return;
                }
                 Lumberjack.error(`Error with tenant key 2 token verification.`, lumberProperties, error2);
               // <INSERT SAME RETRY STUFF>
                // When `exp` claim exists in token claims, jsonwebtoken verifies token expiration.
                throw (error1 instanceof jwt.TokenExpiredError || error2 instanceof jwt.TokenExpiredError)
                    ? new NetworkError(401, "Token expired validated with both key1 and key2.")
                    : new NetworkError(403, "Invalid token validated with both key1 and key2.");
            });
        });
    }
```

Yikes, this code here could probably be cleaned up a lot...

Reply:
I think this has a bit cleaner control flow (and prevents unnecessary key2 check on a 401), but the point is, to make sure we retry if we think the cache was stale before throwing the 403.

```ts

    /**
     * Validates a tenant's API token
     */
    public async validateToken(tenantId: string, token: string, includeDisabledTenant = false, disableCache = false): Promise<void> {
        const tenantKeys = await this.getTenantKeys(tenantId, includeDisabledTenant, disableCache);

        const validateTokenWithKey = async (key: string, keyName: string) => {
            return new Promise<void>((resolve, reject) => {
                jwt.verify(token, key, (error) => {
                    if (!error) {
                        resolve();
                        return;
                    }

                    // When `exp` claim exists in token claims, jsonwebtoken verifies token expiration.
                    reject(error instanceof jwt.TokenExpiredError
                        ? new NetworkError(401, `Token expired validated with ${keyName}.`)
                        : new NetworkError(403, `Invalid token validated with ${keyName}.`))
                });
            });
        };

        // Validate with Key 1
        try {
            await validateTokenWithKey(tenantKeys.key1, KeyName.key1);
        } catch (error) {
            if (isNetworkError(error)) {
                if (error.code === 401 || !tenantKeys.key2) {
                    // Trying key2 with an expired token won't help.
                    // Also, if there is no key2, don't bother validating.
                    throw error;
                }
            }
        }

        // Validate with Key 2
        try {
            await validateTokenWithKey(tenantKeys.key2, KeyName.key2);
        } catch (error) {
            if (isNetworkError(error)) {
                if (error.code === 403 && disableCache === false) {
                    // Assume we used a cached key, and try again by bypassing cache.
                    return this.validateToken(
                        tenantId,
                        token,
                        includeDisabledTenant,
                        true, /* disableCache */
                    );
                }
                throw error;
            }
        }
    }
```

Reply:
@znewton I added a modified version of your changes to ensure that we are invalidating the cache on 403 errors and exiting the validation function once the token is successfully validated